File: reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
/*
思路：
先用fast runner方法，找到中间结点（注意奇偶个数），从中间结点断开，将后段链表翻转，再和前段链表合并。
*/
    void reorderList(ListNode *head) {
        if (!head || head->next == NULL) return;
        
        ListNode dummy(-1);
        dummy.next = head;
        
        // 找到中间结点，slow为中间结点
        ListNode *slow = &dummy, *fast = &dummy;
        while(fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        ListNode *p = slow->next; //后段链表翻转前的头结点
        slow->next = NULL; // 断开前段链表
        
        // 翻转后段链表
        ListNode *head2 = NULL; 
        while(p) {
            ListNode *pn = p->next;
            p->next = head2;
            head2 = p;
            p = pn;
        }
        
        p = &dummy; 
        // 合并两段链表
        while (head && head2) {
            ListNode *hn = head->next;
            ListNode *hn2 = head2->next;
            
            p->next = head;
            head->next = head2;
            p = head2;
            
            head = hn;
            head2 = hn2;
        }
        if (head) { // 奇数个结点时，前段链表多一个结点。
            p->next = head;
        }
        
        head = dummy.next;
        
        return ;
    }
};

File: java-versions/reorder-list.java
/*
143. Reorder List
Medium

2671

133

Add to List

Share
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You may not modify the values in the list's nodes, only nodes itself may be changed.

Example 1:

Given 1->2->3->4, reorder it to 1->4->2->3.
Example 2:

Given 1->2->3->4->5, reorder it to 1->5->2->4->3.
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;
        
        ListNode head2 = null;
        ListNode next;
        while(head != null) {
            next = head.next;
            head.next = head2;
            head2 = head;
            head = next;
        }
        return head2;
    }
    
    private ListNode findMiddleNode(ListNode head) {
        ListNode dummyHead = new ListNode(-1, head);
        ListNode fast = dummyHead;
        ListNode slow = dummyHead;
        while(fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }
    
    private ListNode mergeLists(ListNode left, ListNode right) {
        ListNode dummyHead = new ListNode(-1);
        ListNode t = dummyHead;
        while(right != null) {
            ListNode leftNext = left.next;
            ListNode rightNext = right.next;
            t.next = left;
            t.next.next = right;
            t = right;
            left = leftNext;
            right = rightNext;
        }
        if (left != null) {
            t.next = left;
        }
        return dummyHead.next;
    }
    
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) return;
        
        ListNode middleNode = findMiddleNode(head);
        ListNode right = reverseList(middleNode.next);
        middleNode.next = null; // 断开左右两个链表
        
        head = mergeLists(head, right);
    }
}

